### PR TITLE
Persistence layer implementation based on hibernate ORM

### PR DESCRIPTION
**Pull request by surabujin**
Split persistence layer on areas with ability to use different storages
for each areas. Implement "history" portion of persistence layer via
hibernate ORM. So now we can use any SQL based (by default MySQL)
storage for it.

### CODE REVIEW

**Review by dpoltavets — APPROVED** [review]

**Review by niksv — APPROVED** [review]

File: src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+    @Override
+    public boolean existsByTaskId(String taskId) {
+        return transactionManager.doInTransaction(() -> findEntityByTaskId(taskId).isPresent());
+    }
+
+    @Override
+    public Optional<org.openkilda.model.history.FlowEvent> findByTaskId(String taskId) {
+        return transactionManager.doInTransaction(
+                () -> findEntityByTaskId(taskId).map(org.openkilda.model.history.FlowEvent::new));
```
Comment (rozdy) [inline src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java:64]: No need to use fqn here.

File: src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            HibernateFlowEventDump dump = new HibernateFlowEventDump();
+            FlowEventDumpCloner.INSTANCE.copy(entry.getData(), dump);
+            entity.addDump(dump);
+        }
+
+        return entity;
+    }
+
+    @Override
+    protected FlowEventData doDetach(org.openkilda.model.history.FlowEvent model, HibernateFlowEvent entity) {
```
Comment (rozdy) [inline src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java:135]: Same here.

File: src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        }
+        if (timeTo != null) {
+            filters.add(builder.lessThan(root.get(HibernateFlowEvent_.eventTime), timeTo));
+        }
+        return filters;
+    }
+
+    private Optional<FlowStatusView> extractStatusUpdates(HibernateFlowEventAction actionEntry) {
+        String action = actionEntry.getAction();
+        if (action.equals(org.openkilda.model.history.FlowEvent.FLOW_DELETED_ACTION)) {
```
Comment (rozdy) [inline src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java:167]: And here.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java [hunk 4]
```diff
@@ -15,12 +15,44 @@
...
+import java.util.Set;
+import java.util.concurrent.Callable;
+
 /**
  * Manager of transaction boundaries.
  */
-public interface TransactionManager {
+@Slf4j
+public class TransactionManager implements Serializable {
+    private static final ThreadLocal<ActiveAdapters> globals = new ThreadLocal<>();
```
Comment (rozdy) [inline src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java:40]: `globals` should be uppercase.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java [hunk 5]
```diff
@@ -42,12 +91,180 @@
...
+    }
+
+    /**
+     * Create retry policy using knowledge about possible transient exceptions produced by specific persistence layer.
+     */
+    public <T> RetryPolicy<T> getDefaultRetryPolicy() {
+        RetryPolicy<T> retryPolicy = new RetryPolicy<T>()
+                .handle(RecoverablePersistenceException.class)
+                .withMaxRetries(transactionRetriesLimit)
+                .onRetry(e -> log.debug("Failure in transaction. Retrying #{}...", e.getAttemptCount(),
```
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java:122]: In current "develop" code it is no debug level. But I agree "info" should be better. Updated.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/spi/PersistenceManagerSupplier.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import java.io.Serializable;
+import java.util.function.Supplier;
+
+public class PersistenceManagerSupplier implements Supplier<PersistenceManager>, Serializable {
+    public static PersistenceManagerSupplier DEFAULT = new PersistenceManagerSupplier(new DummyPersistenceManager());
+
+    private PersistenceManager implementation;
+
+    public PersistenceManagerSupplier() {
+        this(null);
```
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/spi/PersistenceManagerSupplier.java:30]: It is for clearness only.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java [hunk 7]
```diff
@@ -42,12 +91,180 @@
...
+            throw ex.getCause();
+        }
+    }
+
+    @PersistenceContextRequired
+    protected <T> T execute(Callable<T> action) throws Exception {
+        TransactionAdapter adapter = adapterFactory.produce(layout);
+
+        ActiveAdapters activeAdapters = initAdapters(adapter);
+        activeAdapters.activate(adapter);
```
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java:151]: Now when you mentioned it... it is really weird. Renaming `ActiveAdapters` into `Transaction`.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java [hunk 8]
```diff
@@ -42,12 +91,180 @@
...
+        try {
+            T result = action.call();
+            activeAdapters.markSuccess();
+            return result;
+        } catch (Exception ex) {
+            log.debug("Failed transaction in {} area in {}", adapter.getArea(), Thread.currentThread().getName(), ex);
```
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java:157]: Main reason for this message is to highlight order of operations during DB transaction. I.e. from TransactionManager perspective it is really "debug" level. Error itself must be handler by the caller, also the caller is responsibe for correct logging.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java [hunk 9]
```diff
@@ -42,12 +91,180 @@
...
+    }
+
+    private <T, E extends Throwable> Callable<T> callableOf(TransactionCallbackWithoutResult<E> action) {
+        return () -> {
+            action.doInTransaction();
+            return null;
+        };
+    }
+
+    private static final class ActiveAdapters {
```
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java:205]: No. Current design suppose such cases at first, but later they (cases) have been declined. Later code will be cleaned to remove such ability completely.

File: src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/HibernateSessionFactorySupplier.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+import org.hibernate.dialect.MySQLDialect;
+
+import java.io.Serializable;
+import java.util.function.Supplier;
+
+@Slf4j
+public class HibernateSessionFactorySupplier implements Supplier<SessionFactory>, Serializable {
+    private final HibernateConfig hibernateConfig;
+
+    private transient volatile SessionFactory product;
```
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/HibernateSessionFactorySupplier.java:43]: yes. renaming into `sessionFactory`.

File: src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java (hunk 1 above)
Comment (surabujin) [reply src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java:64]: yes. removing. It was required before #4376

File: src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java (hunk 2 above)
Comment (surabujin) [reply src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java:135]: updateting

File: src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java (hunk 3 above)
Comment (surabujin) [reply src-java/kilda-persistence-hibernate/src/main/java/org/openkilda/persistence/hibernate/repositories/HibernateHistoryFlowEventRepository.java:167]: updateting

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java (hunk 4 above)
Comment (surabujin) [reply src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionManager.java:40]: I don't agree to treat it as a constant (also it is private, also out checkstyle do not argue against it), but... let it be all uppercase.

File: src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionArea.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.persistence.tx;
+
+public enum TransactionArea {
```
Comment (sergii-iakovenko) [inline src-java/kilda-persistence-api/src/main/java/org/openkilda/persistence/tx/TransactionArea.java:18]: We've already discussed (as a part of #4311) that TransactionArea / TransactionLayout / PersistenceManagerBase.transactionManagerByArea is far from commonly used approaches. It significantly complicates understanding, usage and maintenance of the persistence layer.
The general feeling is over-engineered solution.

Just to remind - I proposed to forbid intersection of orientdb and sql persistence contexts, and force a developer to specify what kind of persistence context is required in each specific case. This solves the problem with less abstraction entities.

**Pull request merged**